inference-engine/src/mkldnn_plugin/nodes/non_max_suppression.cpp

```diff
@@ -135,110 +241,196 @@ class NonMaxSuppressionImpl: public ExtLayerBase {
+                        currBox.score *= coeff(iou);
+                        if (iou >= iou_threshold) {
+                            box_is_selected = false;
+                            break;
                         }
+                        if (currBox.score <= score_threshold)
+                            break;
+                    }
 
-                        if (box_is_selected) {
-                            scores_vector[io_selection_size] = scores_vector[box_idx];
-                            io_selection_size++;
-                            fb.push_back({ scores_vector[box_idx].first, batch, class_idx, scores_vector[box_idx].second });
+                    currBox.suppress_begin_index = fb.size();
+                    if (box_is_selected) {
+                        if (currBox.score == origScore) {
+                            fb.push_back({ currBox.score, batch_idx, class_idx, currBox.idx });
+                            continue;
+                        }
+                        if (currBox.score > score_threshold) {
+                            sorted_boxes.push(currBox);
                         }
                     }
                 }
             }
+            numFiltBox[batch_idx][class_idx] = fb.size();
+            size_t offset = batch_idx*num_classes*max_output_boxes_per_class + class_idx*max_output_boxes_per_class;
+            cpu_memcpy(filtBoxes.data() + offset, fb.data(), fb.size() * sizeof(filteredBoxes));
+        });
+
+        size_t startOffset = numFiltBox[0][0];
+        for (size_t b = 0; b < numFiltBox.size(); b++) {
+            size_t batchOffset = b*num_classes*max_output_boxes_per_class;
+            for (size_t c = (b == 0 ? 1 : 0); c < numFiltBox[b].size(); c++) {
+                size_t offset = batchOffset + c*max_output_boxes_per_class;
+                cpu_memcpy(filtBoxes.data() + startOffset, filtBoxes.data() + offset,
+                           numFiltBox[b][c] * sizeof(filteredBoxes));
+                startOffset += numFiltBox[b][c];
+            }
         }
+        filtBoxes.resize(startOffset);
 
         if (sort_result_descending) {
-            parallel_sort(fb.begin(), fb.end(), [](const filteredBoxes& l, const filteredBoxes& r) { return l.score > r.score; });
+            std::sort(filtBoxes.begin(), filtBoxes.end(), [](const filteredBoxes& l, const filteredBoxes& r) { return l.score > r.score; });
         }
 
-        int selected_indicesStride = outputs[0]->getTensorDesc().getBlockingDesc().getStrides()[0];
-        int* selected_indicesPtr = selected_indices;
+        size_t validOutputs = std::min(filtBoxes.size(), static_cast<size_t>(outputs[NMS_SELECTEDINDICES]->getTensorDesc().getDims()[0]));
+
+        memset(selected_indices, 0, outputs[NMS_SELECTEDINDICES]->byteSize());
+        if (outputs.size() > 1)
+            memset(selected_scores, 0, outputs[NMS_SELECTEDSCORES]->byteSize());
```

Comment:
Why should we memset outputs with zero? Do we have such requirement in spec?

Reply:
fixed